src/v/config/tests/enterprise_property_test.cc

```diff
@@ -96,6 +100,105 @@ TEST(EnterprisePropertyTest, TestRestriction) {
+            return {
+              &cfg.enterprise_bool,
+              {.default_value = false,
+               .allowed_value = false,
+               .restricted_value = true}};
+        }
+        if constexpr (std::same_as<Property, enum_property<ss::sstring>>) {
+            return {
+              &cfg.enterprise_str_enum,
+              {.default_value = "foo",
+               .allowed_value = "baz",
+               .restricted_value = "bar"}};
+        }
+        if constexpr (std::
+                        same_as<Property, property<std::vector<ss::sstring>>>) {
+            return {
+              &cfg.enterprise_str_vec,
+              {.default_value{},
+               .allowed_value = {"OTHER"},
+               .restricted_value = {"GSSAPI", "OTHER"}}};
+        }
+        if constexpr (std::same_as<Property, property<std::optional<int>>>) {
+            return {
+              &cfg.enterprise_opt_int,
+              {.default_value{0}, .allowed_value{10}, .restricted_value{1010}}};
+        }
+        if constexpr (std::same_as<Property, enum_property<tls_version>>) {
+            return {
+              &cfg.enterprise_enum,
+              {.default_value = tls_version::v1_1,
+               .allowed_value = tls_version::v1_2,
+               .restricted_value = tls_version::v1_3}};
+        }
+    }
+
+    test_config cfg{};
+};
+
+using EnterprisePropertyTypes = ::testing::Types<
+  property<bool>,
+  enum_property<ss::sstring>,
+  property<std::vector<ss::sstring>>,
+  property<std::optional<int>>,
+  enum_property<tls_version>>;
+TYPED_TEST_SUITE(EnterprisePropertyTest, EnterprisePropertyTypes);
+
+TYPED_TEST(EnterprisePropertyTest, SanctionedValues) {
+    auto tc = this->get_test_case();
+    auto& enterprise_conf = *tc.enterprise_conf;
+    const auto [default_value, allowed_value, restricted_value] = tc.ts;
+
+    auto binded = enterprise_conf.sanctioning_bind();
+
+    // default value
+    EXPECT_EQ(enterprise_conf.value(), default_value);
+    EXPECT_EQ(enterprise_conf.value(true), default_value);
+    EXPECT_EQ(enterprise_conf.value(false), default_value);
+
+    EXPECT_EQ(binded(true), default_value);
+    EXPECT_EQ(binded(false), default_value);
+
+    // allowed value
+    enterprise_conf.set_value(allowed_value);
+
+    EXPECT_EQ(enterprise_conf.value(), allowed_value);
+    EXPECT_EQ(enterprise_conf.value(true), allowed_value);
+    EXPECT_EQ(enterprise_conf.value(false), allowed_value);
+
+    EXPECT_EQ(binded(true), allowed_value);
+    EXPECT_EQ(binded(false), allowed_value);
+
+    // sanctioned value
+    enterprise_conf.set_value(restricted_value);
+
+    EXPECT_EQ(enterprise_conf.value(), restricted_value);
+    EXPECT_EQ(enterprise_conf.value(true), default_value);
+    EXPECT_EQ(enterprise_conf.value(false), restricted_value);
+
+    EXPECT_EQ(binded(true), default_value);
+    EXPECT_EQ(binded(false), restricted_value);
+}
+
 TEST(EnterprisePropertyTest, TestTypeName) {
```

Comment:
It was a bit confusing for me at first to understand that the other test here aren't using the `struct EnterprisePropertyTest` and its test cases. It might be clearer to rename the struct to avoid this confusion.

Also, parametrizing the tests based on the type feels like magic to me. I probably would have just defined a `void check_test_case<T>(const test_case&)` method and called it multiple times from a single test to keep it simple, but I definitely like the fact that test results are more granular this way. Pretty cool.

Reply:
> `void check_test_case<T>(const test_case&)` 

The problem of doing something like that is that the test suite will have a hard time giving you nice messages in case the test fails. It gets pretty verbose on the setup to parameterize type-tests, though. I do agree.

> `EnterprisePropertyTest`

I think gtest uses this name as the label for the test group, right? I am not very sure but this is why I reused the same name.

Reply:
> I think gtest uses this name as the label for the test group

Yeah, I think that's right. I think it's okay to use multiple different test groups in the same test binary, though. But it's not a big deal.